### ISSUE TITLE
Performance issues on Arthur's computer

### ISSUE THREAD

**Issue by auriamg** [supertuxkart/stk-code#1988]
Arthur_D is experiencing performance issues. Profiling narrowed it down to the "Env map" GPU phase. In shader IBL.frag, the following line seems to be the culprit :

```
Spec = vec4(.25 * SpecularIBL(normal, eyedir, specval), 1.);
```

The slow part in SpecularIBL appears to be the call to "textureLod".

@vlj: maybe this gives you any idea?

**Comment by hiker** [comment]
I also asked Arthur to replace return clamp(textureLod(probe, sampleDirection, lodval).rgb, 0., 1.); in SpecularIBL.frag  with return 0.5;  which make it fast again. So the textureLod appears to be the reason.

**Comment by vlj** [comment]
How big is the performance issue ?
IIRC Arthur_D has a GT310M, which has likely a 64 bits vram bus - ie very slow bandwidth.
SpecularIBL is a texture fetch on a mipmapped 256x256 cubemap per pixel. On the other hand in a deffered pipeline per pixel texture fetch on big textures (eg screensize big) are used in other places, although they don't use mipmap, so I'm surprised it's slow in this particular spot.

**Comment by Flakebi** [comment]
I heard that it drops to 6 fps (from more than 60) when he removes the skybox. And lowering the size of the skybox (to 16x16 or so) doesn't change that.

**Comment by Arthur-D** [comment]
I have over 90 FPS in Gran Paradiso with hiker's workaround, while untouched shader gives me 6 FPS. And yes, changing the texture to 128x128, or even 32x32 or 16x16 doesn't do any difference. My card has the following specifications:

NVIDIA 250 GTS
128 CUDA Cores
1024 MB VRAM
256-bit memory interface
PCI Express x16 Gen2 bus type

**Comment by Arthur-D** [comment]
Some more information:
NVIDIA driver version 340.76
OS: Xubuntu 14.04 for x86_64

Line 10 in SpecularIBL.frag by default is
```    return clamp(textureLod(probe, sampleDirection, lodval).rgb, 0., 1.);```
and gives the following performance:
![20_island-2015 02 19_23 47 44](https://cloud.githubusercontent.com/assets/5515261/6277888/ad746eac-b892-11e4-8be6-c177117bcfb2.png)

Change line 10 in SpecularIBL.frag to
```   return 0.5;```
gives the following performance, but also a graphical issue:
![20_island-2015 02 19_23 49 07](https://cloud.githubusercontent.com/assets/5515261/6277908/d58b1dd2-b892-11e4-9c1b-2d929a3029e6.png)

**Comment by Arthur-D** [comment]
Graphical settings in both screenshots:
![graphics-settings1](https://cloud.githubusercontent.com/assets/5515261/6278047/f15ba09e-b893-11e4-99fa-53fdd7c4536f.jpg)

**Comment by Arthur-D** [comment]
OK, now with IBL shader workaround, I tried at graphics level 5, everything maxed, high-def textures etc etc, max resolution (full HD) and fullscreen. Same track yields lowest FPS 16, mostly around 20.

**Comment by deveee** [comment]
You have this fps drop in a specific case or you just start STK with
typical settings? I tried hiker's workaround and it doesn't change
performance for me. I get exactly the same fps number.

Did you check it with standard nvidia-331 drivers from Ubuntu repositories?
Newer nvidia drivers can cause some problems on linux. For example bug in
Clutter which causes rendering issues on Gnome. I noticed also fps drop in
pcsx2 with nvidia-343 drivers.
